### ISSUE TITLE
ENH: give user control over selection of spatial index frame in sjoin

### ISSUE THREAD

**Issue by adriangb** [geopandas/geopandas#1464]
As discussed in #1421, it could be advantageous to give users some control over which frame (between `left_df` and `right_df`) is used as the spatial index backend, as this can have a huge impact on `sjoin` performance.

One proposal was a `index` keyword that could have values `left`, `right` or `None` (https://github.com/geopandas/geopandas/pull/1421#issuecomment-631342047, https://github.com/geopandas/geopandas/pull/1421#issuecomment-631578974).

ccing @ljwolf, @martinfleis

**Comment by adriangb** [comment]
With #1463, I think we should re-consider the approach here. I see a couple of options:

1. Give `sjoin` a new argument that forces the current implementation of `_geom_predicate_query` to use either `left_df` or `right_df` as it's spatial index. This is essentially the original proposal.
2. Give `sjoin` a new argument that accepts a callable to replace `_geom_predicate_query` with the same signature.
3. Make all of the `sjoin` sub-functions public and allow users to import them and essentially "make" their own `sjoin`, with as much freedom as they want. We still take care of some tricky bits (like making sure there are not column name collisions as well as the entire `_frame_join` part).

I lean towards 2 or 3.

**Comment by adriangb** [comment]
Any thoughts on this @martinfleis  or @jorisvandenbossche ? I think this is no longer needed, aside from maybe documenting the new modular `sjoin` and making the functions public.

**Comment by martinfleis** [comment]
So if anyone wanted to switch from `right_df.sindex` to `left_df.sindex`, you propose writing custom `_geom_predicate_query` and replace it in `sjoin`?

**Comment by adriangb** [comment]
Right. I think it's simpler to understand the relatively clean and simple implementation that currently exists than us trying to figure out all use cases and cover them.

**Comment by martinfleis** [comment]
That is asking for a page in a forthcoming Advanced Guide. Marking this issue as documentation then.

**Comment by bnaul** [comment]
In the documentation vein: I agree that a recipe for more advanced patterns is a good idea, but maybe it also makes sense to clarify in the docstring that (if I'm reading correctly) `right_df.sindex` is used by default unless the predicate is `within`? I think a simple hint could be enough to help most people avoid accidentally sjoining in a much less performant direction.

**Comment by FenderJazz** [comment]
I haven't really got my head around the modular sjoin, but @adriangb's options 2 and 3 seem quite complicated for an average user, who just wants to provide a performance hint, without learning about sjoin internals and writing code to fit it.

Documentation is good, of course, but I don't think it's true that this issue is resolved by documentation alone.  Swapping the order of the GDFs in the sjoin call has implications for the interpretation of all the other arguments, and for the structure of the result, and could require the user to do a lot of recoding and testing to successfully map 'predicate' and 'how', swap 'lsuffix' and 'rsuffix' etc. and then re-order columns.  It seems a pity to impose this on the userbase, rather than write it once within sjoin!
